CASSANDRA-18737: Do not create sstable files before registering in txn

jacek-lewandowski opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableWriter.java:
```diff
@@ -104,11 +107,21 @@ protected SSTableWriter(Builder<?, ?> builder, LifecycleNewTracker lifecycleNewT
 
-        Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
+        List<Component> existingsComponents = components.stream().filter(c -> descriptor.fileFor(c).exists()).collect(Collectors.toList());
+        if (!existingsComponents.isEmpty())
```
We could use `anyMatch` instead of making a collection during the normal path (if necessary, gathering for the error message).

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriter.java:
```diff
@@ -80,14 +80,23 @@ public class BigTableWriter extends SortedTableWriter<BigFormatPartitionWriter>
+            this.indexWriter = builder.getIndexWriter().get();
+
+            checkNotNull(builder.getRowIndexEntrySerializer());
+            checkNotNull(builder.getIndexWriter());
```
I don't think this is doing the intended check, `this.indexWriter` is the one that should not be null.

jacek-lewandowski replied:
right, my bad, moved checkNotNull but forgot to modify the arg

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriter.java:
```diff
@@ -405,13 +414,13 @@ public MmappedRegionsCache getMmappedRegionsCache()
 
         @Override
-        public SequentialWriter getDataWriter()
+        public LazyCloseableRef<SequentialWriter> getDataWriter()
```
I would switch the interface back to returning the objects, but create them here (if necessary, implemented as calling the reference's `get` method).

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriter.java:
```diff
@@ -439,14 +448,14 @@ protected BigTableWriter buildInternal(LifecycleNewTracker lifecycleNewTracker,
-                                                       getIOOptions().flushCompression);
-                indexWriter = new IndexWriter(this);
-                partitionWriter = new BigFormatPartitionWriter(getSerializationHeader(), dataWriter, descriptor.version, rowIndexEntrySerializer.indexInfoSerializer());
+                dataWriter = LazyCloseableRef.create(() -> DataComponent.buildWriter(descriptor,
```
Could this logic be in the `getDataWriter` method? (Renamed to `openDataWriter` for clarity.)

blambov commented on src/java/org/apache/cassandra/utils/LazyCloseableRef.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.function.Supplier;
+import javax.annotation.concurrent.NotThreadSafe;
+
+public abstract class LazyCloseableRef<T extends AutoCloseable> implements Closeable, Supplier<T>
```
Why do we need this to be closeable? The builders don't manage these resources.

I think changing the getters to `open...` as described above will add clarity and make it less likely for incorrect resource management to be introduced later.

jacek-lewandowski replied:
thanks for the idea, will try and see how it looks

jacek-lewandowski committed:
Do not create sstable files before registering in txn

Refactoring prevents the situation where some sstable components, like
data or index, are created before the new sstable is registered with
lifecycle transaction, which leads to a problem such that there is
a short time when incomplete sstable components are present. At the same
time, no transaction file is created, which leads to the possibility
that the sstable can be recognized as completed by various
transaction-aware listers.

blambov commented on src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableWriter.java:
```diff
@@ -358,25 +315,52 @@ public MmappedRegionsCache getMmappedRegionsCache()
+                                                           getIndexWriter().rowIndexWriter,
+                                                           descriptor.version);
+
             return ensuringInBuildInternalContext(partitionWriter);
```
This call is no longer necessary.

blambov commented on src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableWriter.java:
```diff
@@ -358,25 +315,52 @@ public MmappedRegionsCache getMmappedRegionsCache()
         private <T> T ensuringInBuildInternalContext(T value)
         {
-            Preconditions.checkState(value != null, "This getter can be used only during the lifetime of the sstable constructor. Do not use it directly.");
+            checkState(value != null, "This getter can be used only during the lifetime of the sstable constructor. Do not use it directly.");
```
It appears that we will most likely get this if `openPartitionWriter` is called before `openDataWriter` or `openIndexWriter`. Perhaps change the message to say it?

blambov commented on src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableWriter.java:
```diff
@@ -358,25 +315,52 @@ public MmappedRegionsCache getMmappedRegionsCache()
 
         @Override
-        public BtiFormatPartitionWriter getPartitionWriter()
+        protected BtiFormatPartitionWriter openPartitionWriter()
```
I see the problem now... it appears to me that it would be safer to pass index and data writers as arguments, but leave as is if you prefer.

blambov commented on src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableWriter.java:
```diff
@@ -358,25 +315,52 @@ public MmappedRegionsCache getMmappedRegionsCache()
+        protected BtiFormatPartitionWriter openPartitionWriter()
         {
+            checkState(partitionWriter == null, "Partition writer has been already opened.");
+            partitionWriter = new BtiFormatPartitionWriter(getSerializationHeader(),
```
Note: Generally, if the only purpose of this field is to forbid second opening, I would replace it with a boolean, so that there is no chance that this reference causes the garbage collector to keep the object.

This should not be a real problem here as the builder reference is not something that is kept around.

jacek-lewandowski replied:
no, not only that - it is also to close the resources if the constructor fails. And, also to access the already opened resource.

jacek-lewandowski commented:
https://app.circleci.com/pipelines/github/jacek-lewandowski/cassandra/895/workflows/bb8f2560-4e21-442d-8073-cd0865e60e00

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableWriter.java:
```diff
@@ -102,9 +108,38 @@ protected SSTableWriter(Builder<?, ?> builder, LifecycleNewTracker lifecycleNewT
+        // sstable files were created before the sstable is registered in the lifecycle transaction, which may lead
+        // to a race such that the sstable is listed as completed due to the lack of the transaction file before
+        // anything is actually written to it.
+        Supplier<List<Component>> existingComponents = Suppliers.memoize(() -> components.stream()
```
Nit:
I think I gave you the wrong impression what I was looking for here. We are normally working with assertions enabled, so optimizing for the other case is not something I consider important; I was wondering about the amount of work we do in the case where the assertion is checked but does not fire, and this change is making that case worse.

To me it is preferable to not materialize the resulting set, e.g. by doing:
```
        Set<Component> existingComponents = Sets.filter(components, c -> descriptor.fileFor(c).exists());

        assert existingComponents.isEmpty() : String.format("Cannot create a new SSTable in directory %s as a files for components %s already exist there", 
                                                            descriptor.directory, 
                                                            existingComponents);
```
where we create only the filter function and set wrapper at the expense of having to walk and filter the set twice when the assertion fires.

jacek-lewandowski replied:
I was under incorrect impression that we run without assertions enabled

blambov reviewed (approved):
LGTM

mike-tr-adamson commented on src/java/org/apache/cassandra/io/sstable/format/SSTableWriter.java:
```diff
@@ -460,26 +515,17 @@ private static Set<Component> indexComponents(Collection<Index.Group> indexGroup
+            if (this.indexGroups == null)
+                this.indexGroups = ImmutableList.copyOf(indexGroups);
+            else
+                this.indexGroups = ImmutableList.copyOf(Iterables.concat(this.indexGroups, indexGroups));
```
I'm not sure why we need to do this. I can't see any cases where `addSecondaryIndexGroups` is called twice on a builder. This could be potentially dangerous as we could end up calling the same index group twice (or more).

mike-tr-adamson commented on src/java/org/apache/cassandra/io/sstable/format/SSTableWriter.java:
```diff
@@ -98,11 +107,63 @@ protected SSTableWriter(Builder<?, ?> builder, LifecycleNewTracker lifecycleNewT
         this.lifecycleNewTracker = lifecycleNewTracker;
 
+        ArrayList<SSTableFlushObserver> observers = new ArrayList<>();
+        this.observers = Collections.unmodifiableList(observers);
```
nit: Couldn't these 2 lines move down to above the `try` below that populates the observers. I find reading the code that flicking up and down is painful.

mike-tr-adamson commented on src/java/org/apache/cassandra/io/sstable/format/SSTableWriter.java:
```diff
@@ -98,11 +107,63 @@ protected SSTableWriter(Builder<?, ?> builder, LifecycleNewTracker lifecycleNewT
+        // anything is actually written to it.
+        Set<Component> existingComponents = Sets.filter(components, c -> descriptor.fileFor(c).exists());
+
+        assert existingComponents.isEmpty() : String.format("Cannot create a new SSTable in directory %s as a files for components %s already exist there",
```
nit: `Cannot create a new SSTable in directory %s as component files %s already exist there` would read better.

jacek-lewandowski commented:
https://app.circleci.com/pipelines/github/jacek-lewandowski/cassandra/911/workflows/7a876be0-ab14-4800-b19c-3506e831417f
